annotationBatch_EVA-335

poggio84 opened the pull request:
Implementation of the  the steps that annotates the variants in the database using data from a VEP file. 
Using the reader+processor+writer model from Spring Batch.

Commit:
Implementation of the variant annotation load steps using reader processor and writer - EVA-335

jmmut committed:
annotation: improving VEP error logging

jmmut committed:
annotation: fixed property of name of DB collection

jmmut committed:
annotation bugfix: item reader reads compressed files properly

jmmut committed:
annotationLoadBatch: completed tests

Also, bufix of a hardcoded collection name in VariantAnnotationMongoItemWriter

jmmut committed:
annotation: using dynamic flow allows skipping steps

The AnnotBatch job is built using a flow for each step.
Each flow uses a decider to skip or include the step.
The decider has a pipelineOptions to retrieve whether it must
skip the step or not.

The VariantJob still uses the old annotation steps. There's a dedicated
"variantAnnotBatchJob", and the usual skip* work with the new job as well:

skipAnnotGenerateInput=false
skipAnnotCreate=true
skipAnnotLoad=true
will only call the generateInputBatch step

jmmut commented:
I added some things to this PR:
- Improved Vep error logging. Before, if the annotationCreate step failed, the error description was lost, forcing rerunning vep by hand.
- The DB name and collection names are now properly handled, even for mongo with authentication DB. The credentials should be put into user's `application-production.properties`. Putting any of the properties `dbHosts`, `dbAuthenticationDb`, `dbUser` or `dbPassword` in the regular `application.properties` won't have effect because of [Spring's priorities](http://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html): jar-internal `application-production.properties` has higher priority than external `application.propertes`.
- The annotationLoadBatch step was reading the annotation as plain text. I added a `GzipLazyResource` to read it from compressed files. I don't have time to fix it, (that's why a test is failing in travis) but I just noticed that in the annotationGenerateInput step we are writing down the list of variants to annotate in plain text as well. Unfortunately, the same approach with `GzipLazyResource` won't work because spring gets the `File` (not the `OutputStream` from it and instantiates its own `FileInputStream` like this: 

```
File file = FlatFileItemWriter.this.resource.getFile();
os = new FileOutputStream(file.getAbsolutePath(), true);
```
- Completed the tests for the step annotationLoadBatch.
- Added dynamic workflow for the annotationBatch job, in order to be able to skip some steps, but the solution is not perfect. I added an OptionalDecider that queries the corresponding `skip*` from the (singleton) pipelineOptions. Although it works from CLI, I don't know how to do tests similar to `VariantConfigurationTest` because each test needs to configure some parameters (such as which steps to skip, or the `vepInput` property),  and the job gets autowired **before** the code in the tests can change those parameters. It then seems that only with a placeholderConfigurer for each test method it is possble to test the annotationBatch job with skipped steps.

poggio84 commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantAnnotLoadBatchTest.java:
```diff
@@ -24,17 +47,127 @@
+        variantAnnotLoadBatch = new VariantAnnotLoadBatch();
+    }
+
+    @Test
```
This should be moved in a dedicated test for VariantAnnotationLineMapper. The test is for annotation load step, the check if the fields are parsed correctly should be done somewhere else.

poggio84 commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantAnnotLoadBatchTest.java:
```diff
@@ -24,17 +47,127 @@
     @Test
-    public void variantAnnotationReaderShouldReadAllFieldsInVepOutput(){
+    public void variantAnnotationReaderShouldReadAllLinesInVepOutput() throws Exception {
+        String vepOutput = VariantAnnotLoadBatchTest.class.getResource("/annot.tsv.gz").getFile();
```
Why you are using the file and not the vepOutput at the bottom of the test? Using the String you are not depending on external file and is specific for the test + easy to read (no need to open a file)
What about something like:        

```
        ByteArrayOutputStream baostream = new ByteArrayOutputStream();
        OutputStream outStream = new GZIPOutputStream(baostream);
        outStream.write(vepOutput.getBytes("UTF-8"));
        outStream.close();
        byte[] compressedBytes = baostream.toByteArray();


        Resource r = new InputStreamResource(new GZIPInputStream(new ByteArrayInputStream(compressedBytes)));
        FlatFileItemReader<VariantAnnotation> reader = variantsAnnotLoadBatch.initReader(r);
```

poggio84 commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantAnnotLoadBatchTest.java:
```diff
@@ -24,17 +47,127 @@
+                consequenceTypeCount++;
+            }
+        }
+        // all should have at least consequence type annotations
```
Why are we testing this here? Should not be tested during annotation creation step? Here we just reading the annotation that we suppose to be correct. The test should be for checking that all lines are read (and not to check that the lines are correct==confusing)
if some lines are to be skipped then we need to create a processor: from doc.. One typical use for an item processor is to filter out records before they are passed to the ItemWriter. Filtering is an action distinct from skipping; skipping indicates that a record is invalid whereas filtering simply indicates that a record should not be written.

poggio84 commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantAnnotLoadBatchTest.java:
```diff
@@ -24,17 +47,127 @@
+        JobTestUtils.cleanDBs(dbName);
+    }
+
+    private String buildStorageId(String chromosome, int start, String reference, String alternate) {
```
We should definitely move this into some common util class

poggio84 commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/VariantJobsArgs.java:
```diff
@@ -58,9 +57,17 @@
+    @Value("${dbPassword:}") private String dbPassword;
+    @Value("${dbName}") private String dbName;
+    @Value("${dbCollectionVariantsName}") private String dbCollectionVariantsName;
+    @Value("${dbCollectionFilesName}") private String dbCollectionFilesName;
```
Is dbCollectionFilesName used anywhere?

poggio84 commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/VariantJobsArgs.java:
```diff
@@ -58,9 +57,17 @@
+    @Value("${dbUser:}") private String dbUser;
+    @Value("${dbPassword:}") private String dbPassword;
+    @Value("${dbName}") private String dbName;
+    @Value("${dbCollectionVariantsName}") private String dbCollectionVariantsName;
```
Is dbCollectionVariantsName different from dbName?

poggio84 commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/GzipLazyResource.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @author Jose Miguel Mut Lopez &lt;jmmut@ebi.ac.uk&gt;
+ */
+public class GzipLazyResource extends FileSystemResource implements Resource {
```
Is implement Resource necessary? It is already in FileSystemResource AbstractResource

cyenyxe replied:
After checking the hierarchy again, it doesn't seem necessary.

Commit:
Javadoc added

Commit:
Removed

Commit:
Annotation Job IT and configuration

Commit:
dbAuthenticationDb is now optional

Commit:
dbAuthenticationDb is now optional

Commit:
Added suggestion to avoid the usage of an external file

Commit:
Reverted at the pre-dynamic flow state

Commit:
Variant list for now are stored in a non-zipped mode. We will need an extra step to zip the file or a custom FFitemWriter

Commit:
allowStartIfComplete option added

cyenyxe commented:
A couple of general comments:
- readPreference should be set to "primary" in _application.properties_ and to "secondary" in _application-production.properties_ to satisfy recommendations from DBA.
- One of the motivations to rewrite the annotation process was that ELOAD-58 was failing due to an `SOTermNotAvailableException` (see https://github.com/opencb/biodata/commit/a199fa65aca2c42a30c9a7848ed35d07fd53394b), which should be captured and the process prevented from failing. These checks will need to be added at some point, but in that case we will to create a new, specific ticket, to try to merge this one ASAP.

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/ConnectionHelper.java:
```diff
@@ -0,0 +1,74 @@
+ *
+ * @author Jose Miguel Mut Lopez &lt;jmmut@ebi.ac.uk&gt;
+ */
+public class ConnectionHelper {
```
This class should go into a 'utils' package or similar, not in the root of the pipeline.

poggio84 replied:
Should be inside 'pipeline'? Or one level up?

cyenyxe replied:
I don't have a strong preference. In the long term it could end up in the variation-commons repository, because it is such a generic thing. So maybe using an `eva.utils` package instead of `eva.pipeline.utils` would make this transition easier?

poggio84 replied:
Will move in `eva.utils`

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/ConnectionHelper.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class ConnectionHelper {
+    public static List<ServerAddress> parseServerAddresses(String hosts) throws UnknownHostException {
+        List<ServerAddress> erverAddresses = new LinkedList<>();
```
Variable name typo, should be "serverAddresses".

poggio84 replied:
Will fix in next PR

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/ConnectionHelper.java:
```diff
@@ -0,0 +1,74 @@
+        );
+    }
+
+    public static MongoTemplate getMongoTemplate(String database) throws UnknownHostException {
```
Isn't it a bit weird that the two interfaces for this method do not share the first argument?

poggio84 replied:
Fix in next PR

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/OptionalDecider.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author Jose Miguel Mut Lopez &lt;jmmut@ebi.ac.uk&gt;
+ */
+public class OptionalDecider implements JobExecutionDecider {
```
Does it make sense for this class to be in the "annotation" package? Seems generic enough to be in the root pipeline package.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/OptionalDecider.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Override
+    public FlowExecutionStatus decide(JobExecution jobExecution, StepExecution stepExecution) {
```
INFO level log messages shouldn't be so attached to implementation details such as "set to true":
- Skipping step because {} is enabled
- Running step because {} is disabled

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/generateInput/VariantAnnotationItemProcessor.java:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2015-2016 EMBL - European Bioinformatics Institute
```
Copyright should be 2016 only (applies to other files too). @poggio84, @jmmut, do you know if it is acceptable to include the copyright declaration along with the license, to avoid adding all these headers?

poggio84 replied:
Fixed all files in next PR. I've added a copyright profile as described here: https://www.jetbrains.com/help/idea/2016.2/generating-and-updating-copyright-notice.html
Like this copyright is added automatically when you create a new file

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+     *
+     * Most of the code is from org.opencb.biodata.formats.annotation.io.VepFormatReader#read() with few differences:
+     *  - An empty array is initialized for Hgvs (like ConsequenceTypes);
+     *  - parseFrequencies is always true and the all line is always parsed;
```
Could the fact that the whole line is parsed allow to load more fields in the future?

poggio84 replied:
Yes lineFilelds will then contain more element to accommodate in VariantAnnotation

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+        // Initialize Hgvs
+        if(currentAnnotation.getHgvs()==null) {
+            currentAnnotation.setHgvs(new ArrayList<>());
+        }
```
Are these 2 null checks necessary? It seems that they will always be null at this point?

poggio84 replied:
Removed both

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+
+        /**
+         * parses extra column and populates fields as required. Some lines do not have extra field and end with a \t: the split function above does not return that field
+         * true parameter indicates the function to also parse frequencies
```
I think the comment "true parameter indicates the function to also parse frequencies" does not apply any more since there is no reference to the parameter in the code

poggio84 replied:
Removed

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+         * parses extra column and populates fields as required. Some lines do not have extra field and end with a \t: the split function above does not return that field
+         * true parameter indicates the function to also parse frequencies
+         */
+        if(lineFields.length>13) {
```
`lineFields.length>13` or always exactly 14?

poggio84 replied:
I don't know :( in this case I just copied the logic from org.opencb.biodata.formats.annotation.io.VepFormatReader#read()
How can I check this? Or to whom should I ask?

cyenyxe replied:
There are 2 options: checking the VEP output documentation in the Ensembl website, or asking the original author, but he might not remember anyway XD

poggio84 replied:
The default output format ("VEP" format when downloading from the web interface) is a 14 column tab-delimited file. The output columns are:
Moved to 14

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+     * From org.opencb.biodata.formats.annotation.io.VepFormatReader
+     * #parseRemainingFields(org.opencb.biodata.models.variant.annotation.ConsequenceType, java.lang.String[])
+     */
+    private void parseRemainingFields(ConsequenceType consequenceType, String[] lineFields) {
```
This method could be renamed to `parseTranscriptFields()` for clearer description of its task. Field assignment doesn't follow order found (6th index is the last to assign).

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+        } else {
+            consequenceType.setSoTermsFromSoNames(Arrays.asList(lineFields[6].split(",")));   // fill so terms
+        }
+        currentAnnotation.getConsequenceTypes().add(consequenceType);
```
I assume VariantAnnotation doesn't have a method `addConsequenceType()`? :sob:

poggio84 replied:
Only setConsequenceTypes(List<ConsequenceType> consequenceTypes) :(

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+//            parsedVariant.put("chromosome", variantLocationFields[0]);
+//            parsedVariant.put("start", variantLocationFields[1]);
+            parsedVariant.put("end", (variantLocationFields.length > 2) ? variantLocationFields[2] : variantLocationFields[1]);
+        } catch (ArrayIndexOutOfBoundsException e) {
```
The process shouldn't abort but flag the record as not valid.

poggio84 replied:
I can remove it but this method is a exact copy/paste of org.opencb.biodata.formats.annotation.io.VepFormatReader#parseVariant
so should be changed there as well?

cyenyxe replied:
Looks like the last version of the biodata library throws the exception instead of exiting, so don't bother back-porting it to older versions.

poggio84 replied:
Done

cyenyxe replied:
Shouldn't the exception be thrown for Spring Batch to detect it is an incorrect entry? (same for line 181)

poggio84 replied:
Should we skip the line (annotation) if malformed? If yes then I agree we should tell spring to skip it

cyenyxe replied:
Yes, the line should be skipped but the step should keep running fine for the rest. Is it possible to write failed records to a specific output, for our own tracking purposes or potential reporting to Ensembl?

poggio84 replied:
Done in next commit

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationLineMapper.java:
```diff
@@ -0,0 +1,376 @@
+            parsedVariant.put("start", leftVariantFields[leftVariantFields.length-2]);
+            parsedVariant.put("reference", leftVariantFields[leftVariantFields.length-1]);
+            parsedVariant.put("alternative", variantFields[1]);
+        } catch (ArrayIndexOutOfBoundsException e) {
```
The process shouldn't abort but flag the record as not valid.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationMongoItemWriter.java:
```diff
@@ -0,0 +1,168 @@
+            throw new RuntimeException("Unable to initialize MongoDB", e);
+        }
+        return mongoTemplate;
+    }*/
```
Please remove commented code block.

poggio84 replied:
OK

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationMongoItemWriter.java:
```diff
@@ -0,0 +1,168 @@
+    @Override
+    protected void doWrite(List<? extends VariantAnnotation> variantAnnotations) {
+        //// TODO: 30/06/2016 move in constructor?
+        DBObjectToVariantAnnotationConverter converter = new DBObjectToVariantAnnotationConverter();
```
Looks like [it doesn't have any internal state](https://github.com/opencb/opencga/blob/hotfix/0.5/opencga-storage/opencga-storage-mongodb/src/main/java/org/opencb/opencga/storage/mongodb/variant/DBObjectToVariantAnnotationConverter.java), so it should be fine.

poggio84 replied:
OK moved in constructor

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationMongoItemWriter.java:
```diff
@@ -0,0 +1,168 @@
+        //// TODO: 30/06/2016 move in constructor?
+        DBObjectToVariantAnnotationConverter converter = new DBObjectToVariantAnnotationConverter();
+
+        for (VariantAnnotation variantAnnotation : variantAnnotations) {
```
Could VariantAnnotation objects be grouped by chr+start+ref+alt at the beginning of the method, before starting the loop, in order to reduce the number of queries to the database?

poggio84 replied:
In case of two annotations (or more) in the same variant?

cyenyxe replied:
Yes.

poggio84 replied:
And PRE-merging the VariantAnnotation before the query... Yes! I will look into this

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationMongoItemWriter.java:
```diff
@@ -0,0 +1,168 @@
+                    .findOne(find, new BasicDBObject("annot", 1));
+
+            //update annotation in existing variant
+            if (null != existingVariantStorage){
```
`variable != null` instead of `null != variable` to follow the same code style than the rest.

poggio84 replied:
OK will swap.. I was just following this:
https://en.wikipedia.org/wiki/Yoda_conditions
http://stackoverflow.com/questions/271561/why-does-one-often-see-null-variable-instead-of-variable-null-in-c

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationMongoItemWriter.java:
```diff
@@ -0,0 +1,168 @@
+                    //update ConsequenceTypes
+                    VariantAnnotation existingAnnotation = converter.convertToDataModelType((DBObject) existingVariantStorage.get("annot"));
+
+                    if(null != variantAnnotation.getConsequenceTypes()){
```
`variable != null` instead of `null != variable` to follow the same code style than the rest.

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/annotation/load/VariantAnnotationMongoItemWriter.java:
```diff
@@ -0,0 +1,168 @@
+
+                    //update Hgvs
+                    if(null != variantAnnotation.getHgvs()){
+                        if(null == existingAnnotation.getHgvs()){
```
`variable != null` (or equals) instead of `null != variable` to follow the same code style than the rest.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/jobs/VariantAnnotConfiguration.java:
```diff
@@ -31,10 +31,12 @@
 @Configuration
 @EnableBatchProcessing
+@Import(VariantJobArgsConfig.class)
 public class VariantAnnotConfiguration {
```
Should be overwritten with `VariantAnnotaConfigurationBatch.java`.

poggio84 replied:
Maybe we can deprecate this class until we switch to the new annotation (batch) job?

cyenyxe replied:
Oh, I thought the new annotation job was already integrated into the VariantConfiguration class but now I see it's not... Ideally we should test it with a real file and have it merged by the time this ticket is closed. Would you see any reason to postpone it?

poggio84 replied:
Unless you need to skip some step independently then yes can be integrated into VariantConfiguration. (Btw this will change again when we will add the Dynamic Flow)

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/jobs/VariantAnnotConfigurationBatch.java:
```diff
@@ -0,0 +1,101 @@
+    @Autowired private Step variantAnnotLoadBatchStep;
+
+    @Qualifier("annotationCreate")
+    @Autowired private Step annotationCreate;
```
Do you think autowiring could be reduced a bit? It doesn't seem really necessary to me here (please correct me otherwise), and would make things less obvious when we start encapsulating more steps in other classes.

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/jobs/VariantAnnotConfigurationBatch.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Bean
+    @Qualifier("annotationCreate")
+    public Step annotationCreate() {
```
This step should be extracted the same way as the others, even if is still is a tasklet instead of a reader-processor-writer.

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatch.java:
```diff
@@ -0,0 +1,190 @@
+@Configuration
+@EnableBatchProcessing
+@Import(VariantJobArgsConfig.class)
+public class VariantsAnnotGenerateInputBatch {
```
Merge into `VariantsAnnotGenerateInput.java`

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatch.java:
```diff
@@ -0,0 +1,190 @@
+                .processor(processor)
+                .writer(writer).allowStartIfComplete(false)
+                .build();
+    }*/
```
Remove commented code block.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatch.java:
```diff
@@ -0,0 +1,190 @@
+
+    @Bean
+    public ItemReader<DBObject> variantReader() throws Exception {
+        return initReader(pipelineOptions.getString("dbCollectionVariantsName"), mongoOperations());
```
Why extracting reader and writer initialization to extra (also public) methods?

poggio84 replied:
I was using them in test.. that's why but yes I agree not very nice to have. I removed the public method and re-worked a bit the test as well

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatch.java:
```diff
@@ -0,0 +1,190 @@
+            );
+        }
+        return mongoTemplate;
+    }
```
`mongoOperations()` and `getMongoTemplate()` methods present here as well as in `ConnectionHelper`. Should they be extracted (and the class renamed to `MongoConnectionHelper`)?

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatch.java:
```diff
@@ -0,0 +1,190 @@
+@EnableBatchProcessing
+@Import(VariantJobArgsConfig.class)
+public class VariantsAnnotGenerateInputBatch {
+
```
How hard is to make a class inherit from Step directly? I've seen documentation on how to do it using XML, but not Java annotations.

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotLoadBatch.java:
```diff
@@ -0,0 +1,142 @@
+@Configuration
+@EnableBatchProcessing
+@Import(VariantJobArgsConfig.class)
+public class VariantsAnnotLoadBatch {
```
Merge with `VariantsAnnotLoad`

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotLoadBatch.java:
```diff
@@ -0,0 +1,142 @@
+        return jobs.get(jobName)
+                .start(variantAnnotLoadBatchStep())
+                .build();
+    }*/
```
Remove commented code block

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/main/resources/application-production.properties:
```diff
@@ -3,3 +3,10 @@ job.repository.driverClassName=org.postgresql.Driver
 job.repository.url=jdbc:postgresql://10.7.248.20:49153/spring_batch_annotations
 job.repository.username=postgres
 job.repository.password=batch
```
Job repository credentials should be removed and values filled using Maven profiles.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/jobs/AnnotationConfig.java:
```diff
@@ -0,0 +1,26 @@
+ */
+@Configuration
+@PropertySource({"annotation.properties"})
+public class AnnotationConfig {
```
Not sure I understand the purpose of this class.

poggio84 replied:
It is loading the ApplicationContext with props.. otherwise you get:
`java.lang.IllegalStateException: Failed to load ApplicationContext
`

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/jobs/AnnotationConfig.java:
```diff
@@ -0,0 +1,26 @@
+    @Bean
+    public JobLauncherTestUtils jobLauncherTestUtils() {
+                return new JobLauncherTestUtils();
+           }
```
`JobLauncherTestsUtils.class` can be added to the @ContextConfiguration in the JUnit test class, making this `@Bean` method unnecessary.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/jobs/VariantAnnotConfigurationBatchTest.java:
```diff
@@ -0,0 +1,230 @@
+        while (cursor.hasNext()) {
+            cnt++;
+            VariantAnnotation annot = converter.convertToDataModelType((DBObject)cursor.next().get("annot"));
+            TestCase.assertTrue(annot.getConsequenceTypes() != null);
```
Could be `assertNotNull`.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/jobs/VariantAnnotConfigurationBatchTest.java:
```diff
@@ -0,0 +1,230 @@
+
+    private String readLine(File outputFile) throws IOException {
+        if (reader == null) {
+            reader = new BufferedReader(new FileReader(outputFile));
```
Could the BufferedReader be created/destroyed in `@Before` and `@After` methods?

poggio84 replied:
Not in `@Before` because the file doesn't exist yet. I'm using now try with resources so I don't need to close the buffer. But still initialized in the method and not in `@Before`

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/jobs/VariantAnnotConfigurationBatchTest.java:
```diff
@@ -0,0 +1,230 @@
+        return (DBObject) JSON.parse(variant);
+    }
+
+    private final String variantWithoutAnnotation = "{\n" +
```
Extracting to plain file into test/resources folder would simplify making any changes to test input.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatchTest.java:
```diff
@@ -0,0 +1,294 @@
+
+    @Before
+    public void setUp() throws Exception {
+        databaseName = PropertiesLoaderUtils.loadAllProperties("application.properties").getProperty("mongo.db");
```
Could the property for the database name be the same as in the rest of the app? ("dbName" right now)

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotGenerateInputBatchTest.java:
```diff
@@ -0,0 +1,294 @@
+    }
+
+
+    private final String variantWithoutAnnotation = "{\n" +
```
Extracting the 2 variables below to plain file into test/resources folder would simplify making any changes to test input.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotLoadBatchTest.java:
```diff
@@ -0,0 +1,195 @@
+        VariantAnnotationLineMapper lineMapper = new VariantAnnotationLineMapper();
+        for (String annotLine : vepOutput.split("\n")) {
+            VariantAnnotation variantAnnotation = lineMapper.mapLine(annotLine, 0);
+            assertTrue(variantAnnotation.getConsequenceTypes() != null);
```
Could be `assertNotNull`.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotLoadBatchTest.java:
```diff
@@ -0,0 +1,195 @@
+
+        FlatFileItemReader<VariantAnnotation> reader = variantsAnnotLoadBatch.initReader(new GzipLazyResource(vepOutput));
+
+/*      // Suggestion to avoid to use an external file
```
If it is possible and easy not to use an external file, that would be good.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotLoadBatchTest.java:
```diff
@@ -0,0 +1,195 @@
+        JobTestUtils.cleanDBs(dbName);
+    }
+
+    private String buildStorageId(String chromosome, int start, String reference, String alternate) {
```
Will need to think where to place this to avoid repetitions...

poggio84 replied:
What about a `VariantsHelper` class in eva.utils?

cyenyxe replied:
We could even re-use the Mongo?ConnectionHelper since this is a pretty much Mongo-specific thing.

poggio84 replied:
Done

cyenyxe commented on eva-pipeline/src/test/java/embl/ebi/variation/eva/pipeline/steps/VariantsAnnotLoadBatchTest.java:
```diff
@@ -0,0 +1,195 @@
+        return builder.toString();
+    }
+
+    private final String vepOutput = "" +
```
Could be extracted to plain file in test/resources folder.

poggio84 replied:
In this case is it not easier to have this here? Easy to read and no need to open a new file to check the content

Commit:
Moved into utils package; Typo fixed for serverAddresses; Database is now the first arg

Commit:
ConnectionHelper moved into utils package; Database name is now the first arg in getMongoTemplate

Commit:
Moved into pipeline root and INFO message changed

Commit:
Fixed Copyright

Commit:
ConnectionHelper moved one level up in eva.utils

Commit:
Removed unnecessary null checks
Renamed parseRemainingFields to parseTranscriptFields
LineFields are now processed in order in parseTranscriptFields
The process in not aborting in case of ArrayIndexOutOfBoundsException in parseVariant

Commit:
Refactor and general improvements

Commit:
File extracted from class, used in test

Commit:
New utility class to handle mongo bits

Commit:
makeGzipFile method moved here

Commit:
readPreference added

Commit:
readPreference option added

Commit:
jobLauncherTestUtils bean removed

Commit:
Vep output is containing 14 fields.. change in the code accordingly

Commit:
mongoOperations externalized in MongoDBHelper

Commit:
Removed unused import

Commit:
Refactor with general improvements

Commit:
Removed unused import

cyenyxe commented on eva-pipeline/src/main/java/embl/ebi/variation/eva/pipeline/MongoDBHelper.java:
```diff
@@ -0,0 +1,114 @@
+                return ReadPreference.secondary();
+            default:
+                throw new IllegalArgumentException(
+                        String.format("%s is not a valid ReadPreference type, please double check!", readPreference));
```
A string like the following could help the user to solve the problem:

> %s is not a valid ReadPreference type, please use "primary" or "secondary"

poggio84 replied:
Done

Commit:
Removed ad merged into original file

Commit:
Merged from VariantsAnnotLoadBatchTest

Commit:
Removed tasklet and added spring batch reader/processor/writer functionality

Commit:
Reusing VariantAnnotConfigurationTest

Commit:
Removed tasklet and added spring batch reader/processor/writer functionality

Commit:
Removed skip step option for steps no longer valid

Commit:
VariantAnnotConfiguration is now replacing the annotation steps

Commit:
Removed skip step option for steps no longer valid

Commit:
Removed implements Resource as not necessary

Commit:
VariantConfig job is now using the Annotation flow. Test updated.

Commit:
Improved exception message

Commit:
Log improved

Commit:
Merged from VariantsAnnotGenerateInputBatch

Commit:
Reusing VariantAnnotConfigurationBatch

Commit:
readPreference and auth db options added

Commit:
Removed properties details

Commit:
Removed opencga property value

Commit:
Set opencga value to /opt/opencga

Commit:
Trying to fix NPE in StorageManagerFactory.parseStorageEngineName

Commit:
Malformed annotations are skipped and logged

Commit:
Added a SkipListener to log skipped line into a different file

Commit:
VariantAnnotation(s) are pre-merged before to be persisted reducing the number of queries to the database

Commit:
Removed stream and groupingBy because not working in travis
